// @nomicfoundation/hardhat-network-helpers/dist/src/internal/network-helpers/lazy-network-helpers.js
import { bindAllMethods } from "@nomicfoundation/hardhat-utils/lang";
import { Duration } from "./duration/duration.js";
let NetworkHelpersImpl;
export class LazyNetworkHelpers {
    #connection;
    #impl;
    time;
    constructor(connection) {
        this.#connection = connection;
        this.time = new LazyTime(() => this.#getImpl());
        bindAllMethods(this);
    }
    clearSnapshots() {
        // No-op before the impl is constructed — there are no snapshots to clear.
        this.#impl?.clearSnapshots();
    }
    async dropTransaction(txHash) {
        const impl = await this.#getImpl();
        return await impl.dropTransaction(txHash);
    }
    async getStorageAt(address, index, block) {
        const impl = await this.#getImpl();
        return await impl.getStorageAt(address, index, block);
    }
    async impersonateAccount(address) {
        const impl = await this.#getImpl();
        return await impl.impersonateAccount(address);
    }
    async loadFixture(fixture) {
        const impl = await this.#getImpl();
        return await impl.loadFixture(fixture);
    }
    async mine(blocks, options) {
        const impl = await this.#getImpl();
        return await impl.mine(blocks, options);
    }
    async mineUpTo(blockNumber) {
        const impl = await this.#getImpl();
        return await impl.mineUpTo(blockNumber);
    }
    async setBalance(address, balance) {
        const impl = await this.#getImpl();
        return await impl.setBalance(address, balance);
    }
    async setBlockGasLimit(blockGasLimit) {
        const impl = await this.#getImpl();
        return await impl.setBlockGasLimit(blockGasLimit);
    }
    async setCode(address, code) {
        const impl = await this.#getImpl();
        return await impl.setCode(address, code);
    }
    async setCoinbase(address) {
        const impl = await this.#getImpl();
        return await impl.setCoinbase(address);
    }
    async setNextBlockBaseFeePerGas(baseFeePerGas) {
        const impl = await this.#getImpl();
        return await impl.setNextBlockBaseFeePerGas(baseFeePerGas);
    }
    async setNonce(address, nonce) {
        const impl = await this.#getImpl();
        return await impl.setNonce(address, nonce);
    }
    async setPrevRandao(prevRandao) {
        const impl = await this.#getImpl();
        return await impl.setPrevRandao(prevRandao);
    }
    async setStorageAt(address, index, value) {
        const impl = await this.#getImpl();
        return await impl.setStorageAt(address, index, value);
    }
    async stopImpersonatingAccount(address) {
        const impl = await this.#getImpl();
        return await impl.stopImpersonatingAccount(address);
    }
    async takeSnapshot() {
        const impl = await this.#getImpl();
        return await impl.takeSnapshot();
    }
    async #getImpl() {
        if (NetworkHelpersImpl === undefined) {
            ({ NetworkHelpers: NetworkHelpersImpl } = await import("./network-helpers.js"));
        }
        if (this.#impl === undefined) {
            this.#impl = new NetworkHelpersImpl(this.#connection);
        }
        return this.#impl;
    }
}
class LazyTime {
    // Eager: `Duration` is pure math with no heavy imports, so instantiating it
    // up-front doesn't undermine the lazy-loading goal of this wrapper.
    duration = new Duration();
    #getImpl;
    constructor(getImpl) {
        this.#getImpl = getImpl;
        bindAllMethods(this);
    }
    async increase(amountInSeconds) {
        const impl = await this.#getImpl();
        return await impl.time.increase(amountInSeconds);
    }
    async increaseTo(timestamp) {
        const impl = await this.#getImpl();
        return await impl.time.increaseTo(timestamp);
    }
    async latest() {
        const impl = await this.#getImpl();
        return await impl.time.latest();
    }
    async latestBlock() {
        const impl = await this.#getImpl();
        return await impl.time.latestBlock();
    }
    async setNextBlockTimestamp(timestamp) {
        const impl = await this.#getImpl();
        return await impl.time.setNextBlockTimestamp(timestamp);
    }
}
