// @nomicfoundation/hardhat-verify/dist/src/internal/lazy-verification.js
import { bindAllMethods } from "@nomicfoundation/hardhat-utils/lang";
let VerificationImpl;
export class LazyVerification {
    etherscan;
    #provider;
    #networkName;
    #chainDescriptors;
    #verificationProvidersConfig;
    #impl;
    constructor(provider, networkName, chainDescriptors, verificationProvidersConfig) {
        this.#provider = provider;
        this.#networkName = networkName;
        this.#chainDescriptors = chainDescriptors;
        this.#verificationProvidersConfig = verificationProvidersConfig;
        this.etherscan = new LazyEtherscanForwarder(() => this.#getImpl());
        bindAllMethods(this);
    }
    async #getImpl() {
        if (VerificationImpl === undefined) {
            ({ Verification: VerificationImpl } = await import("./verification-helpers.js"));
        }
        if (this.#impl === undefined) {
            this.#impl = new VerificationImpl(this.#provider, this.#networkName, this.#chainDescriptors, this.#verificationProvidersConfig);
        }
        return this.#impl;
    }
}
class LazyEtherscanForwarder {
    #getImpl;
    constructor(getImpl) {
        this.#getImpl = getImpl;
        bindAllMethods(this);
    }
    async getChainId() {
        const impl = await this.#getImpl();
        return await impl.etherscan.getChainId();
    }
    async getName() {
        const impl = await this.#getImpl();
        return await impl.etherscan.getName();
    }
    async getUrl() {
        const impl = await this.#getImpl();
        return await impl.etherscan.getUrl();
    }
    async getApiUrl() {
        const impl = await this.#getImpl();
        return await impl.etherscan.getApiUrl();
    }
    async getApiKey() {
        const impl = await this.#getImpl();
        return await impl.etherscan.getApiKey();
    }
    async getContractUrl(address) {
        const impl = await this.#getImpl();
        return await impl.etherscan.getContractUrl(address);
    }
    async isVerified(address) {
        const impl = await this.#getImpl();
        return await impl.etherscan.isVerified(address);
    }
    async verify(args) {
        const impl = await this.#getImpl();
        return await impl.etherscan.verify(args);
    }
    async pollVerificationStatus(guid, contractAddress, contractName) {
        const impl = await this.#getImpl();
        return await impl.etherscan.pollVerificationStatus(guid, contractAddress, contractName);
    }
    async customApiCall(params, options) {
        const impl = await this.#getImpl();
        return await impl.etherscan.customApiCall(params, options);
    }
}
